### PR TITLE
Feature/1 deregistration

### PR DESCRIPTION
**Pull request by piotr-bugara-gravity9**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -97,6 +97,17 @@ manager.start();
 
 For more info about `fullDocument`, `fullDocumentBeforeChange` and `maxAwaitTime` see https://www.mongodb.com/docs/manual/reference/method/db.collection.watch/
 
+
+### Listener deregistration
+
+Each listener can be deregistred at any moment. It can be done only for specific
```
Comment (piotrlg9) [inline README.md:103]: Spelling: `deregistred` -> `deregistered`

File: src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java [hunk 2]
```diff
@@ -162,10 +162,19 @@ public void run() {
         } while (true);
     }
 
-    void register(ChangeStreamListener listener) {
-        log.info("Registering listener {} to worker on partition {} for collection {}", listener.getClass().getName(), partition, mongoConfig.getCollectionName());
-        listeners.add(listener);
-    }
+	void register(ChangeStreamListener listener) {
+		log.info("Registering listener {} to worker on partition {} for collection {}", listener.getClass().getName(), partition, mongoConfig.getCollectionName());
```
Comment (piotrlg9) [inline src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java:166]: I would replace here `listener.getClass().getName()` with `listener` or `listener.getName()` so client code can implement unique name for logging purposes.

File: src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java [hunk 3]
```diff
@@ -162,10 +162,19 @@ public void run() {
...
+		listeners.add(listener);
+	}
+
+	public void deregister(ChangeStreamListener listener) {
+		log.info("Unregistering listener {} from worker on partition {} for collection {}", listener.getClass().getName(), partition, mongoConfig.getCollectionName());
+		listeners.remove(listener);
```
Comment (piotrlg9) [inline src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java:172]: To support this line `listeners` must be a synchronized collection.

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java [hunk 4]
```diff
@@ -19,100 +21,174 @@
...
+        assertEquals(1, deleteResult0.getDeletedCount());
+        assertEquals(1, deleteResult1.getDeletedCount());
+        assertEquals(1, deleteResult2.getDeletedCount());
+
+        // Wait for CDC event to arrive
+        Thread.sleep(500);
+
+        // only old events are available
+        List<ChangeStreamDocument<Document>> events1 = listener0.getEvents();
+        assertEquals(3, events1.size());
```
Comment (piotrlg9) [inline src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:139]: 3 should be replaced here by events0.size()

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java [hunk 5]
```diff
@@ -19,100 +21,174 @@
...
+        assertEquals(1, deleteResult0.getDeletedCount());
+        assertEquals(1, deleteResult1.getDeletedCount());
+        assertEquals(1, deleteResult2.getDeletedCount());
+
+        // Wait for CDC event to arrive
+        Thread.sleep(500);
+
+        // only old events and deletes from not deregistered partitions are available
+        List<ChangeStreamDocument<Document>> events1 = listener0.getEvents();
+        assertEquals(5, events1.size());
```
Comment (piotrlg9) [inline src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:174]: 5 is a magic number here which depends on:
* events0.size
* number of deleteOne calls
* size of list used for deregisterListener call (more or less)

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java [hunk 6]
```diff
@@ -19,100 +21,174 @@
...
+    }
+
+    private void insertDocumentsToAllPartitions() {
```
Comment (piotrlg9) [inline src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:177]: Should return number of events generated and the value should be used in asserts.

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java [hunk 7]
```diff
@@ -19,100 +21,174 @@
...
+                .collectionName(getTestCollectionName())
+                .workerConfigCollectionName(getWorkerConfigCollectionName())
+                .clusterConfigCollectionName(getClusterConfigCollectionName())
+                .numberOfPartitions(3)
+                .fullDocument(FullDocument.UPDATE_LOOKUP)
+                .build();
+        collection = MongoClientProvider.createClient(getConnectionUri()).getDatabase(getDatabaseName()).getCollection(getTestCollectionName());
+    }
+
+    private MongoConfig mongoConfig;
```
Comment (piotrlg9) [inline src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:45]: Ordering - should be moved up.

**Review by piotrlg9 — CHANGES_REQUESTED** [review]

File: src/main/java/com/gravity9/mongocdc/MongoCDCManager.java [hunk 8]
```diff
@@ -48,55 +49,78 @@ private Map<Integer, MongoChangeStreamWorker> createWorkers() {
...
+            var workerOptional = getMongoChangeStreamWorker(listener, partition);
+            workerOptional.ifPresent(worker -> deregister(worker, listener, partition));
+        });
+    }
+
+    public void deregisterListenerFromAllPartitions(ChangeStreamListener listener) {
+        workers.values().forEach(worker -> worker.deregister(listener));
+    }
+
+    private void deregister(MongoChangeStreamWorker worker, ChangeStreamListener listener, int partition) {
```
Comment (piotrlg9) [inline src/main/java/com/gravity9/mongocdc/MongoCDCManager.java:110]: You can put all this code inside the worker - [Tell Don't Ask](https://martinfowler.com/bliki/TellDontAsk.html) and return `boolean` if needed (= `false` if listener not found).

Comment (piotr-bugara-gravity9) [reply src/main/java/com/gravity9/mongocdc/MongoCDCManager.java:110]: done. Return statement not added - i think it's not needed

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java (hunk 7 above)
Comment (piotr-bugara-gravity9) [reply src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:45]: done

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java (hunk 6 above)
Comment (piotr-bugara-gravity9) [reply src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:177]: done

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java (hunk 5 above)
Comment (piotr-bugara-gravity9) [reply src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:174]: that's why there is a comment
change a little bit. Is it more clear now?

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java (hunk 4 above)
Comment (piotr-bugara-gravity9) [reply src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java:139]: done

File: README.md (hunk 1 above)
Comment (piotr-bugara-gravity9) [reply README.md:103]: done

File: src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java (hunk 3 above)
Comment (piotr-bugara-gravity9) [reply src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java:172]: done. Used `CopyOnWriteArraySet` as you proposed

File: src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java (hunk 2 above)
Comment (piotr-bugara-gravity9) [reply src/main/java/com/gravity9/mongocdc/MongoChangeStreamWorker.java]: Right now - only toString is called. We have to think on how to implement it better

File: src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java (hunk 5 above)
Comment (piotrlg9) [reply src/test/java/com/gravity9/mongocdc/ChangeStreamTest.java]: LGTM

**Review by piotrlg9 — APPROVED** [review]

**Pull request merged**